components/eventing-controller/controllers/subscription/nats/unit_test.go

```diff
@@ -0,0 +1,309 @@
+package nats
+
+import (
+	"context"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	kymalogger "github.com/kyma-project/kyma/common/logging/logger"
+	eventingv1alpha1 "github.com/kyma-project/kyma/components/eventing-controller/api/v1alpha1"
+	"github.com/kyma-project/kyma/components/eventing-controller/logger"
+	"github.com/kyma-project/kyma/components/eventing-controller/pkg/handlers/mocks"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func TestHandleSubscriptionDeletion(t *testing.T) {
+	g := NewGomegaWithT(t)
+	testEnvironment := setupTestEnvironment(t)
+	ctx, r, mockedBackend := testEnvironment.Context, testEnvironment.Reconciler, testEnvironment.Backend
+
+	subscription := &eventingv1alpha1.Subscription{
+		ObjectMeta: v1.ObjectMeta{
+			Name:       "test",
+			Namespace:  namespaceName,
+			Finalizers: []string{}, // empty finalizers
+		},
+	}
+
+	err := testEnvironment.Reconciler.Client.Create(testEnvironment.Context, subscription)
+	g.Expect(err).Should(BeNil())
+
+	err = r.handleSubscriptionDeletion(ctx, subscription, r.namedLogger())
+	g.Expect(err).Should(BeNil())
+
+	// the subscription has no kyma eventing finalizers and hence should not be processed in the function
+	mockedBackend.AssertNotCalled(t, "DeleteSubscription", subscription)
+	g.Expect(subscription.ObjectMeta.Finalizers).Should(BeEmpty())
+	g.Expect(err).Should(BeNil())
+
+	// add the eventing finalizer to test the function's branches
+	subscription.ObjectMeta.Finalizers = []string{"eventing.kyma-project.io"}
+	mockedBackend.On("DeleteSubscription", subscription).Return(nil)
+	err = r.handleSubscriptionDeletion(ctx, subscription, r.namedLogger())
+	g.Expect(err).Should(BeNil())
+
+	// the function should have cleared the finalizers
+	mockedBackend.AssertCalled(t, "DeleteSubscription", subscription)
+	g.Expect(subscription.ObjectMeta.Finalizers).Should(BeEmpty())
+
+	// check the changes were made on the kubernetes server
```

Comment:
do we need to do it in unit tests as well? Isn't already covered in the integration tests?

Reply:
Perhaps we already have integration tests for that that functionality, but the idea of this ticket were to write the unit tests and tests all the possible branches of the code. So, that's why it's here. IMO having more tests (integration/unit) is never bad. And the integration could do it maybe on a higher level.